### PR TITLE
feat: [CON-1372] Implement the `ArtifactAssembler` for `DownloadStrippedArtifactMessage`

### PR DESCRIPTION
**Pull request by kpop-dfinity**
Implements the `ArtifactAssembler` trait together with fetching missing ingress messages logic.

What's missing (will be part of subsequent PRs):
1. Stripping ingress messages from block propols
2. Metrics

### CODE REVIEW

**Comment by kpop-dfinity** [comment]
Hey @tthebst can you take a look at this before I mark the PR as ready? It's still missing the `Stripped` types definitions, I'll add it in the next PR, if it's okay

File: rs/types/types/src/batch/ingress.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ use std::{
 #[cfg_attr(test, derive(ExhaustiveSet))]
 pub struct IngressPayload {
     /// Pairs of MessageId and its serialized byte position in the buffer.
-    id_and_pos: Vec<(IngressMessageId, u64)>,
+    pub id_and_pos: Vec<(IngressMessageId, u64)>,
```
Comment (tthebst) [inline rs/types/types/src/batch/ingress.rs:20]: With `get_by_id` this does not need to be `pub`.

File: rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            ));
+        }
+
+        while let Some(join_result) = join_set.join_next().await {
+            let Ok((ingress, _peer_id)) = join_result else {
+                return Err(Aborted {});
+            };
+
+            stripped_block_proposal
+                .try_insert_ingress_message(ingress)
```
Comment (tthebst) [inline rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs:177]: Shouldn't this and `try_assemble` be infallible?

File: rs/p2p/artifact_downloader/src/fetch_stripped_artifact/download.rs [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+        }
+    }
+
+    #[tokio::test]
+    async fn rpc_get_from_ingress_pool_test() {
+        let ingress_message = SignedIngressBuilder::new().nonce(1).build();
+        let block = fake_block_proposal(vec![]);
+        let pools = mock_pools(Some(ingress_message.clone()), None);
+
+        let rpc_response = rpc_handler(
```
Comment (tthebst) [inline rs/p2p/artifact_downloader/src/fetch_stripped_artifact/download.rs:249]: I would have directly tested on the axum server here with `build_axum_router` and `Oneshot(request)`. This tests the more public interface and also catches issues with the router config.

Comment (kpop-dfinity) [reply rs/p2p/artifact_downloader/src/fetch_stripped_artifact/download.rs]: Good idea, done

File: rs/types/types/src/batch/ingress.rs [hunk 4]
```diff
@@ -100,6 +101,22 @@ impl IngressPayload {
         self.id_and_pos.is_empty()
     }
 
+    pub fn get_by_id(
+        &self,
+        ingress_message_id: &IngressMessageId,
+    ) -> Result<SignedIngress, IngressPayloadError> {
```
[the remark this answers is not in the document]
Comment (kpop-dfinity) [reply rs/types/types/src/batch/ingress.rs]: Agreed, done

File: rs/protobuf/src/gen/p2p/registry.crypto.v1.rs [hunk 5]
```diff
@@ -0,0 +1,210 @@
+/// A list of subnets that can sign with this ECDSA key.
```
Comment (rumenov) [inline rs/protobuf/src/gen/p2p/registry.crypto.v1.rs:1]: this seems unrelated

Comment (tthebst) [inline rs/protobuf/src/gen/p2p/registry.crypto.v1.rs:1]: Since we now have 2k lines of duplicate generated protobuf code it might indeed be better to put the type next to the consensus protobufs.

**Review by tthebst — APPROVED** [review]

File: rs/protobuf/src/gen/p2p/registry.crypto.v1.rs (hunk 5 above)
Comment (tthebst) [reply rs/protobuf/src/gen/p2p/registry.crypto.v1.rs:1]: The new protos use protos defined in proto `types` and it regenerates all proto dependencies. If we would put the new `Stripped` type in `def/types` we would not need to regenerate all.

**Comment by rumenov** [comment]
There is a slight problem with the MR. Now the http handler can put unboundedly messages into the unvalidated pool.

Maybe consider incrementing the peer counter when a messages is inserted in any of the pools and decrement the counter when an ingress message is completely removed from the pools

**Review by rumenov — APPROVED** [review]

**Review by DSharifi — APPROVED** [review]

File: rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    async fn assemble_message<P: Peers + Clone + Send + 'static>(
+        &self,
+        id: <MaybeStrippedConsensusMessage as IdentifiableArtifact>::Id,
+        artifact: Option<(MaybeStrippedConsensusMessage, NodeId)>,
+        peer_rx: P,
+    ) -> Result<(ConsensusMessage, NodeId), Aborted> {
+        // Download the Stripped message if it hasn't been pushed.
+        let (stripped_artifact, peer) = self
+            .fetch_stripped
+            .assemble_message(id.clone(), artifact, peer_rx.clone())
```
Comment (eichhorl) [inline rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs:142]: Maybe I'm confused, why does the function `assemble_message` return a `stripped_artifact`?

**Review by eichhorl — APPROVED** [review]

File: rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs (hunk 6 above)
Comment (tthebst) [reply rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs:142]: Since stripped artifacts may still get advertised it can happen that you only receive the `StrippedMessageId`. So the first step is to download the `StrippedMessage` by doing an rpc. But yes, the naming could be better.

Comment (eichhorl) [reply rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs:143]: I see, thanks!
